spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/NetChangelogIterator.java

```diff
@@ -0,0 +1,138 @@
+import org.apache.spark.sql.types.StructType;
+
+public class NetChangelogIterator extends ChangelogIterator {
+
+  private Row firstRow = null;
+  private Row postImage = null;
+
+  protected NetChangelogIterator(
+      Iterator<Row> rowIterator, StructType rowType, String[] identifierFields) {
+    super(rowIterator, rowType, identifierFields);
+  }
+
+  @Override
+  public boolean hasNext() {
+    if (postImage != null || firstRow != null) {
+      return true;
+    }
+    return getRowIterator().hasNext();
+  }
+
+  /**
+   * This method iterates all versions of the same logical row across multiple snapshots. It caches
+   * the row's first version, and removes every version in between, then compares the last version
+   * with the first version to get the net change. Please note that it also removes any invalid
+   * data. For example, if there are two identical records, it will remove one of them.
+   */
+  @Override
+  public Row next() {
+    // return the cached row if there is one
+    if (postImage != null) {
+      Row currentRow = postImage;
+      postImage = null;
+      return currentRow;
+    }
+
+    // return the first row if there is no more rows
+    if (!getRowIterator().hasNext() && firstRow != null) {
+      Row currentRow = firstRow;
+      firstRow = null;
+      return currentRow;
+    }
+
+    if (firstRow == null) {
+      firstRow = getRowIterator().next();
+      return null;
+    }
+
+    Row lastRow = null;
+
+    while (getRowIterator().hasNext()) {
+      Row currentRow = getRowIterator().next();
+      // if current row is not the same logical as the first row
+      if (sameLogicalRow(firstRow, currentRow)) {
+        lastRow = currentRow;
+      } else {
+        // if current row is not the same logical row as the first row, compute the results and
+        // set the current row as the first row, and return the result
+        Row result = getNetChanges(firstRow, lastRow);
+        firstRow = currentRow;
+        return result;
+      }
+    }
+
+    Row current = getNetChanges(firstRow, lastRow);
+    firstRow = null;
+    return current;
+  }
+
+  /**
+   * This method gets the net changes between the first version and the last version of a logical
+   * row. Here are the algorithm:
+   *
+   * <ol>
+   *   <li>If the first version is an insert, and last one is an insert, then return an insert with
+   *       the last version.
+   *   <li>If the first version is an insert, and last one is a delete, then return null.
+   *   <li>If the first version is a delete, and last one is a delete, then return a delete.
+   *   <li>If the first version is a delete, and last one is an insert, if they are the exactly
+   *       same, they are carry-over rows, return null, otherwise return an update pair.
+   * </ol>
+   */
+  private Row getNetChanges(Row first, Row last) {
+    if (first == last || last == null) {
+      return first;
+    }
+
+    if (first.getString(getChangeTypeIndex()).equals(INSERT)) {
```

Comment:
Could we move `getChangeTypeIndex` to a separate variable?

Reply:
It'd be nice to reuse it. Open for suggestion though.